src/core/qgsvectorfilewriter.cpp

```diff
@@ -2749,21 +2750,27 @@ QString QgsVectorFileWriter::driverForExtension( const QString &extension )
   QString ext = extension.trimmed();
   if ( ext.isEmpty() )
     return QString();
-  QgsApplication::registerOgrDrivers();
-  int const drvCount = OGRGetDriverCount();
+
+  if ( ext.startsWith( '.' ) )
+    ext.remove( 0, 1 );
+
+  GDALAllRegister();
+  int const drvCount = GDALGetDriverCount();
 
   for ( int i = 0; i < drvCount; ++i )
   {
-    OGRSFDriverH drv = OGRGetDriver( i );
+    GDALDriverH drv = GDALGetDriver( i );
     if ( drv )
     {
-      QString drvName = OGR_Dr_GetName( drv );
-      if ( OGR_Dr_TestCapability( drv, "CreateDataSource" ) != 0 )
+      char **driverMetadata = GDALGetMetadata( drv, nullptr );
+      if ( CSLFetchBoolean( driverMetadata, GDAL_DCAP_CREATE, false ) && CSLFetchBoolean( driverMetadata, GDAL_DCAP_VECTOR, false ) )
       {
-        MetaData metadata;
-        if ( driverMetadata( drvName, metadata ) )
+        QString drvName = GDALGetDriverShortName( drv );
+        QStringList driverExtensions = QString( GDALGetMetadataItem( drv, GDAL_DMD_EXTENSIONS, nullptr ) ).split( ' ' );
```

Comment:
Does QString() accept a nullptr without segfaulting ? If not, it might be prudent testing the return of GDALGetMetadataItem() in case of a drive that would not declare extensions but still have create + vector capabilities.